File: src/InstaxFrame.cpp

/**
Instax Client Library
By: Scott McKittrick

InstaxFrame - Class for building and managing data packets sent between client and printer

*/

#include "InstaxFrame.hpp"

namespace InstaxClient
{
  /*
    The frame is built by inserting one byte at a time. 
    Each call to this function takes a byte. It will return true when a full packet is recieved
    It will throw an exception when something is wrong with the frame.
  */
  bool InstaxFrame::recvPacket(uint8_t byte)
  {
    readQueue.push_back(byte);

    //If we haven't recieved the full header yet
    if(readQueue.size() < PACKET_DATA_OFFSET)
      return false;

    //If we have recieved the full header, let's parse it
    if(readQueue.size() == PACKET_DATA_OFFSET)
      {
	parseHeader();
	return false;
      }

    if(readQueue.size() == length)
      {

	//Verify the checksum
	if(verifyChecksum())
	  {
	    return true;
	  }
	else
	  {
	    throw InvalidFrameException("Invalid frame checksum");
	  }  
      }

    //if we get here, we must be waiting on more data bytes
    return false;
  }

  /*
  The Instax Printer uses a modified TCP Checksum
  Essentially it is a normal TCP checksum but instead of a 16 bit value, they use a 32 bit value
  With the original 16 bit value as the upper two bytes and two static numbers as the lower two.
  */
  const bool InstaxFrame::verifyChecksum()
  {
    uint16_t cksum = 0;
    int checksumOffset = readQueue.size() - PACKET_CHECKSUM_LENGTH; 
    //Start by checking the magic numbers
    if(readQueue[checksumOffset + 2] != PACKET_CHECKSUM_MAGIC1)
      {
	return false;
      }

    if(readQueue[checksumOffset + 3] != PACKET_CHECKSUM_MAGIC2)
      {
	return false;
      }
    
    //Start by adding all data bytes together 
    for(int i = 0; i < checksumOffset; i++)
      {
	cksum += readQueue[i];
      }

    //Now we put in the recieved checksum and the result should equal 0xFFFF
    cksum += (readQueue[checksumOffset] << 8) | readQueue[checksumOffset + 1];
    return cksum == 0xFFFF;
    
  }

  //takes a data array and calculates the checksum
  uint32_t InstaxFrame::generateChecksum(const std::vector<uint8_t>& packet)
  {
    uint16_t calcSum = 0;
    for(int i = 0; i < packet.size(); i++)
      {
	calcSum += packet[i];
      }

    uint32_t result = ~calcSum << 16;

    //Add in the magic numbers
    result |= PACKET_CHECKSUM_MAGIC1 << 8;
    return result | PACKET_CHECKSUM_MAGIC2;
  }

  void InstaxFrame::parseHeader()
  {
    //Parse out Values
    SID = readQueue[PACKET_SID_OFFSET];

    length = readQueue[PACKET_LENGTH_OFFSET] << 8 | readQueue[PACKET_LENGTH_OFFSET + 1];

    UID = readQueue[PACKET_UID_OFFSET] << 24;
    UID |= readQueue[PACKET_UID_OFFSET + 1] << 16;
    UID |= readQueue[PACKET_UID_OFFSET + 2] << 8;
    UID |= readQueue[PACKET_UID_OFFSET + 3];

    password = readQueue[PACKET_PASSWORD_OFFSET] << 8 | readQueue[PACKET_PASSWORD_OFFSET +1];
  }

  const uint8_t InstaxFrame::getSID()
  {
    return SID;
  }

  const uint16_t InstaxFrame::getLength()
  {
    return length;
  }

  const uint32_t InstaxFrame::getUID()
  {
    return UID;
  }

  const uint16_t InstaxFrame::getPassword()
  {
    return password;
  }

  //Copys the packet data into the provided vector
  //Returns the length of the copied data
  const int InstaxFrame::getPacketData(std::vector<uint8_t> * dest)
  {
    //Check that there is data to copy
    int dataLen = length - PACKET_DATA_OFFSET - PACKET_CHECKSUM_LENGTH;
    if(dataLen == 0)
      //there is no data in the packet
      return 0;

    std::vector<uint8_t>::iterator dataStart = readQueue.begin() + PACKET_DATA_OFFSET;
    std::vector<uint8_t>::iterator dataEnd = readQueue.end() - PACKET_CHECKSUM_LENGTH;

    dest->assign(dataStart, dataEnd);
    return dataLen;
  }

      /*
      Build a frame ready to be sent over the network.
      Takes a pointer to the vector the packet should be added
      Takes packet parameters and vector of data
    */
  void InstaxFrame::buildOutgoingPacket(std::vector<uint8_t> * packet,
                                    const uint8_t startByte,
                                    const uint8_t SID,
                                    const uint32_t UID,
                                    const uint16_t password,
                                    const std::vector<uint8_t>& data)
  {
    int packetLength = data.size() + PACKET_DATA_OFFSET + PACKET_CHECKSUM_LENGTH;

    //ToDo Handle packets that are too long for the 2 byte packet length value. Throw an exception
    
    packet->push_back(startByte);
    
    packet->push_back(SID);
    
    packet->push_back((packetLength >> 8) & 0x00FF);
    packet->push_back(packetLength & 0x00FF);

    packet->push_back((UID >> 24) & 0xFF);
    packet->push_back((UID >> 16) & 0xFF);
    packet->push_back((UID >> 8) & 0xFF);
    packet->push_back(UID & 0xFF);

    packet->push_back((password >> 8) & 0xFF);
    packet->push_back(password & 0xFF);

    packet->push_back(0x00);
    packet->push_back(0x00);
    
    packet->insert(packet->end(), data.begin(), data.end());

    uint32_t cksum = generateChecksum(*packet);
    packet->push_back((cksum >> 24) & 0xFF);
    packet->push_back((cksum >> 16) & 0xFF);
    packet->push_back((cksum >> 8) & 0xFF);
    packet->push_back(cksum & 0xFF);
  }
    
}

File: research/emulator/InstaxFrame.java
import java.util.ArrayList;

public class InstaxFrame
{
	//Packet Definitions
	protected static final int PACKET_SID_OFFSET = 1;
	protected static final int PACKET_LENGTH_OFFSET = 2;
	protected static final int PACKET_UID_OFFSET = 4;
	protected static final int PACKET_PASSWORD_OFFSET = 8;
	protected static final int PACKET_DATA_OFFSET = 12;
	protected static final int PACKET_HEADER_LENGTH = 12;
	protected static final int PACKET_CHECKSUM_LENGTH = 4;
	protected static final int magicNumber1 = 13;
	protected static final int magicNumber2 = 10;
	
	public static final int CLIENT_FRAME_START = 0x24;
	public static final int PRINTER_FRAME_START = 0x2A;
	public static final short CLIENT_DEFAULT_PASSWORD = (short)0xFFFF;
	public static final short PRINTER_DEFAULT_PASSWORD = (short)0x0000;
	
	protected static final int startingArrLength = 20;
	
	//Packet Arrays
	protected ArrayList<Integer> incommingArray;
	
	//Frame header data
	protected int SID; //Command id
	protected int packetLength; //Length of whole packet including header
	protected int UID; //Device id
	protected short password;
	
	public InstaxFrame()
	{
		incommingArray = new ArrayList<Integer>(startingArrLength);
	}
	
	public static int[] buildOutgoingPacket(int startByte, int SID, int UID, short password, int[] dataArr)
	{
		int[] data;
		if(dataArr == null)
			data = new int[0];
		else
			data = dataArr;
		
		int packetLength = data.length + PACKET_CHECKSUM_LENGTH + PACKET_HEADER_LENGTH;
		int[] outgoingArray = new int[packetLength];
		int pointer = 0;
		//Packet start value
		outgoingArray[pointer++] = startByte;
		//Command id
		outgoingArray[pointer++] = SID;
		//Packet length
		outgoingArray[pointer++] = 0xFF & (packetLength >> 8);
		outgoingArray[pointer++] = 0xFF & (packetLength);
		//device id
		outgoingArray[pointer++] = 0xFF & (UID >> 24);
		outgoingArray[pointer++] = 0xFF & (UID >> 16);
		outgoingArray[pointer++] = 0xFF & (UID >> 8);
		outgoingArray[pointer++] = 0xFF & (UID);
		//Password
		outgoingArray[pointer++] = 0xFF & (password >> 8);
		outgoingArray[pointer++] = 0xFF & (password);
		//Header end bytes
		outgoingArray[pointer++] = 0x00;
		outgoingArray[pointer++] = 0x00;
		
		//Write the data
		System.arraycopy(data, 0, outgoingArray, pointer, data.length);
		pointer += data.length;
		
		//Calculate the checksum
		int[] cksum = calculateChecksum(outgoingArray, packetLength - 4);
		outgoingArray[pointer++] = cksum[0];
		outgoingArray[pointer++] = cksum[1];
		outgoingArray[pointer++] = cksum[2];
		outgoingArray[pointer++] = cksum[3];
		return outgoingArray;
	}
	
	//We will read the packet one byte at a time for simplicity on an untrusted byte stream.
	public boolean recvPacket(int inByte) throws InvalidFrameException
	{
		incommingArray.add(inByte);
		//If we have the header, parse it.
		if(incommingArray.size() == PACKET_HEADER_LENGTH)
			parseHeader();
		
		//If we haven't received the full header yet, we haven't gotten the full packet.
		if(incommingArray.size() < PACKET_HEADER_LENGTH)
			return false;
		
		//If we have received the entire packet
		if(incommingArray.size() == packetLength)
		{
			if(verifyChecksum(incommingArray.toArray(new Integer[0]), incommingArray.size() - 4))
				return true;
			else
				throw new InvalidFrameException("Checksum verification failed");
		}
		
		//If we have made it this far, we haven't gotten the packet yet.
		return false;
	}
	
	protected void parseHeader() throws InvalidFrameException
	{
		//System.out.println("Header received.");
		if(incommingArray.size() < PACKET_HEADER_LENGTH)
			throw new InvalidFrameException("Packet is shorter than header length");
			
		SID = incommingArray.get(PACKET_SID_OFFSET);
		packetLength = (0xFF & incommingArray.get(PACKET_LENGTH_OFFSET)) << 8  | (0xFF & (int)incommingArray.get(PACKET_LENGTH_OFFSET + 1));
		//System.out.println("Packet Length: " + packetLength);
		UID = (int)(0xFF & incommingArray.get(PACKET_UID_OFFSET)) << 24 |
			  (int)(0xFF & incommingArray.get(PACKET_UID_OFFSET + 1)) << 16 |
			  (int)(0xFF & incommingArray.get(PACKET_UID_OFFSET + 2)) << 8 |
			  (int)(0xFF & incommingArray.get(PACKET_UID_OFFSET + 3));
		password = (short)((0xFF & incommingArray.get(PACKET_PASSWORD_OFFSET)) << 8  | (0xFF & (int)incommingArray.get(PACKET_PASSWORD_OFFSET + 1)));
	}
	
	protected static int[] calculateChecksum(final int[] array, final int n)
	{
		int i = 0;
		int cksum = 0;
		int[] cksumBytes = new int[4];
		while(i < n)
		{
			cksum += (0xFF & array[i]);
			i++;
		}
		cksumBytes[0] = (int)(0xFF & ~cksum >> 8);
		cksumBytes[1] = (int)(0xFF & ~cksum >> 0);
		cksumBytes[2] = 13;
		cksumBytes[3] = 10;
		return cksumBytes;
	}
	
	protected static boolean verifyChecksum(final Integer[] array, int n)
	{
		int cksum = 0;
		for(int i = 0; i < n; i++)
		{
			cksum += (0xFF & array[i]);
		}
		//System.out.println("Calculated: " + Integer.toHexString(cksum));
		return (0xFFFF & cksum + ((0xFF & array[n]) << 8 | (0xFF & array[n+1])<<0)) == 0xFFFF && array[n+2] == magicNumber1 && array[n+3] == magicNumber2;
	}
	
	public int getSID()
	{
		return SID;
	}
	
	public int getUID()
	{
		return UID;
	}
	
	public short getPassword()
	{
		return password;
	}
	
	public String toString()
	{
		StringBuilder b = new StringBuilder();
		b.append("InstaxFrame Dump: \n");
		b.append("SID: " + Integer.toHexString(SID) + "\n");
		b.append("UID: " + Integer.toHexString(UID) + "\n");
		b.append("Data: " );
		for(int i = 0; i < incommingArray.size(); i++)
		{
			b.append(Integer.toHexString(incommingArray.get(i)) + " ");
		}
		b.append("\n");
		return b.toString();
	}
	
	public int getDataByte(int n)
	{
		return incommingArray.get(PACKET_HEADER_LENGTH + n);
	}

	public int getShort(int n)
	{
		return (incommingArray.get(PACKET_HEADER_LENGTH + n) & 0xFF)<<8 | (incommingArray.get(PACKET_HEADER_LENGTH + n + 1)&0xFF);
	}
	
	public int getInt(int n)
	{
		return (incommingArray.get(PACKET_HEADER_LENGTH + n) & 0xFF << 24) | (incommingArray.get(PACKET_HEADER_LENGTH + n + 1) & 0xFF << 16) | (incommingArray.get(PACKET_HEADER_LENGTH + n + 2) & 0xFF << 8) | (incommingArray.get(PACKET_HEADER_LENGTH + n + 3) & 0xFF << 0);
	}
	
	public int[] getDataArray()
	{
		int dataLength = packetLength - PACKET_HEADER_LENGTH - PACKET_CHECKSUM_LENGTH;
		System.out.println("PacketLength: " + packetLength);
		System.out.println("DataLength: " + dataLength);
		int[] data = new int[dataLength];
		for(int i = 0; i < dataLength; i++)
			data[i] = incommingArray.get(PACKET_HEADER_LENGTH + i);
		return data;
	}
}